config/checker-framework-suppressions/checker-nullness-optional-interning-suppressions.xml

```diff
@@ -5358,6 +5358,13 @@
     </details>
   </checkerFrameworkError>
 
+  <checkerFrameworkError unstable="false">
+    <fileName>src/main/java/com/puppycrawl/tools/checkstyle/grammar/CompositeLexerContextCache.java</fileName>
+    <specifier>dereference.of.nullable</specifier>
+    <message>dereference of possibly-null reference currentContext</message>
+    <lineContent>if (currentContext.getCurlyBraceDepth() == 0) {</lineContent>
+  </checkerFrameworkError>
```

Comment:
Tried a few things to avoid this, but there was always some conflict with checker/pitest/coverage (when adding `null` checks), so I went with the least "bad" of the three that also provided the best code (the most important point).

The issue is that `Deque#peek` is nullable. Since we check to see if the stack is empty before we enter the code here, we will never hit this issue. In the case that we ever did get a NPE exception here (impossible as the code is now), some exception would be desirable because someone would have introduced a bug anyway.

Also, we can't push `null` to a Deque without throwing a NPE either.

I went with the code that makes the most sense from a readability/maintainability standpoint.

`Deque#peek != null` is not as idiomatic as `!Deque.isEmpty`, period. The Collections framework provides `isEmpty`, and we should use it.

Battle with checker:

```
  <checkerFrameworkError unstable="false">
    <fileName>src/main/java/com/puppycrawl/tools/checkstyle/grammar/CompositeLexerContextCache.java</fileName>
    <specifier>introduce.eliminate</specifier>
    <message>It is bad style to create an Optional just to chain methods to get a non-optional value.</message>
    <lineContent>.orElseThrow(</lineContent>
  </checkerFrameworkError>
```

```
  <checkerFrameworkError unstable="false">
    <fileName>src/main/java/com/puppycrawl/tools/checkstyle/grammar/CompositeLexerContextCache.java</fileName>
    <specifier>unboxing.of.nullable</specifier>
    <message>unboxing a possibly-null reference curlyBraceCounterStack.peek()</message>
    <lineContent>if (curlyBraceCounterStack.peek() == 0) {</lineContent>
  </checkerFrameworkError>

```

Battle with pitest and code coverage:
```
[INFO] --- jacoco-maven-plugin:0.8.11:check (default-check) @ checkstyle ---
[INFO] Loading execution data file /home/vsts/work/1/s/target/jacoco.exec
[INFO] Analyzed bundle 'checkstyle' with 833 classes
[WARNING] Rule violated for class com.puppycrawl.tools.checkstyle.grammar.CompositeLexerContextCache: branches covered ratio is 0.90, but expected minimum is 1.00
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  05:00 min
[INFO] Finished at: 2024-03-02T16:22:07Z
[INFO] ------------------------------------------------------------------------

```

Reply:
`private final Deque<TemplateContext> templateContextStack = new ArrayDeque<>();`
do we have tool that do not like this code ?

Reply:
>do we have tool that do not like this code ?

No, the issue is explained in detail above. the issue is that `peek` is never null, but some tools want a null check. I went with the most readable/idiomatic code that did not allow any pitest mutations to survive.